raft-tech/TANF-app
# Proof of Concept TDP Based Kibana Auth

## Summary of Changes
- Added routes to browse to Kibana from the frontend
- Added a very generic/simple authentication check view to backend to redirect to Kibana if user has correct permissions.
Pull request closes #1858 

## Considerations
- Implementing auth in this manner prevents us from leveraging Elastic/Kibana RBAC. All users who can access Kibana will be able to read/write any index in Elastic. In the future, if we wanted granular RBAC, we will need to write our own plugin for Elastic to integrate our auth with their realm model. Maybe there are open source options that exist for this? We could also theoretically make direct requests to the Elastic server security API to try and hack RBAC by creating the user/group in realtime when we authenticate from the backend. 
- When deployed, we will need to deny any external incoming traffic on the port we deploy the Elastic/Kibana server on to prevent users from stepping around TDP auth and accessing Elastic/Kibana directly. Maybe there is also a way to do this with NGINX if all traffic is routed through that.
- Is there a way we can get around the redirect/hide the Kibana URL and make it seem TPD native?

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up --build
cd tdrs-backend && docker-compose up --build
```

- Open http://localhost:3000.
- Browse to /kibana before logging in and verify that you cannot see Kibana/stay at the login page
- Sign in as an admin/data analyst/developer user
- Browse to /kibana or click the navbar item. Verify that you still cant see kibana because you did not authenticate with hhs ams.
- Add `BYPASS_KIBANA_AUTH="yes"` to your backend `.env` file, browse to /kibana and verify you can see the Kibana home page.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Dev-ready requirements for #1350 and potential follow-on tickets
- [x] Recommended cloudfoundry security guidelines and configurations
- [x] TDP user logic and flows for authorization
- [x] Kibana/ES permissions and configurations guidelines/recommendations
- [ ] Application security hardening both for Kibana/ES and TDP
- [ ] Investigate/automate security scans around our data pool

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)


### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
